### ISSUE TITLE
404 Errors when Panning and Zooming

### ISSUE THREAD

**Issue by HelmerX88** [interactive-game-maps/grand_theft_auto_san_andreas#1]
Hello,

I've noticed there are a bunch of 404 Issues with this map when zooming out and moving the map, for example:

       GET https://interactive-game-maps.github.io/grand_theft_auto_san_andreas/map_tiles/2/-12/4.png 404 (Not Found)
       GET https://interactive-game-maps.github.io/grand_theft_auto_san_andreas/map_tiles/2/-10/-6.png 404 (Not Found)

This occours also with some other maps, like the one for GTA IV.
I believe I found a potential fix trying some stuff, I'm not sure though, since I dont fully understand this part of the Code:

```
// takes array of coords [x,y(,z)]
// returns L.latLng
function gtaCoordinatesToLeaflet(coords) {
    // The gta map 0,0 is in the middle, 6000 square
    // 192 (our map) / 6000 (gta map) = 0.032
    lx = (coords[0] + 3000) * 0.032;
    ly = (coords[1] - 3000) * 0.032;
    return L.latLng(ly, lx);
}
```

I can't figure out why **192** is used, where does this number come from exactly?

For GTA IV its:

```
var lx = (coords[0] + 2571) * 0.0321;
var ly = (coords[1] - 2441) * 0.03187;
```

Are **2571** and **2441** the middle of the map for GTA IV?
Also why do we use **0.0321** here, isn't the map also 6000 wide?

Excuse me if I'm missing something obvious here, I haven't really worked with Leaflet before.

**Comment by interactive-game-maps** [comment]
Hi!

> I've noticed there are a bunch of 404 Issues with this map when zooming out and moving the map

> I believe I found a potential fix trying some stuff

AFAIK this is a quirk of leaflet itself, it also happens on maps that don't convert coordinates like the aporia one.
According to some issues on the leaflet tracker it can be avoided using the [`bounds`](https://leafletjs.com/reference.html#tilelayer-bounds) option for tile layers.

> I can't figure out why **192** is used, where does this number come from exactly?

This is the size of the map in leaflets coordinate system itself. Our target value.
https://interactive-game-maps.github.io/grand_theft_auto_san_andreas/?share=192,-192

> Are **2571** and **2441** the middle of the map for GTA IV?
> Also why do we use **0.0321** here, isn't the map also 6000 wide?

I don't remember exactly. The images are 6000x3800. With the same formula of SA we get

~~~
x1 = 6000 * 0,0321 = 192.6
x2 = 3800 * 0,03187 = 121.106
~~~

It's close: https://interactive-game-maps.github.io/grand_theft_auto_iv/?share=192.6,-121.106
But not quite: https://interactive-game-maps.github.io/grand_theft_auto_iv/?share=187.5,-118.75

I remember that the scaling of the maps were weird and I had to manually align and crop them to overlay each other correctly (ingame view and wireframe view). I should have documented that better.
It might be that the image 0,0 doesn't align with the ingame 0,0, which would explain the shifting not being half of the image size.
I honestly don't know anymore, sorry.

Actually, I think I slightly remember. GTA 4 was particularly obnoxious because the ingame 0,0 isn't in the middle of the map.

Great question btw, you got me :)

**Comment by HelmerX88** [comment]
Hello, thanks for the reply.

> According to some issues on the leaflet tracker it can be avoided using the [bounds](https://leafletjs.com/reference.html#tilelayer-bounds) option for tile layers.

That's what I found to be working for the SA map, I've added

```
bounds:[
             gtaCoordinatesToLeaflet([-3000, -3000]),
             gtaCoordinatesToLeaflet([3000, 3000]) 
],
```

to the TileLayer in **map.js** and it seems to render all the tiles at all zoom levels properly without the 404 errors.

However, I couldn't get this working for the GTA IV one, mainly because I couldn't figure out the max X and Y values for it, due to the weird scaling I guess.

> > I can't figure out why 192 is used, where does this number come from exactly?
> 
> This is the size of the map in leaflets coordinate system itself. Our target value.

So does Leaflet calculate that value based on the tilesize at a certain zoom level, is it defined somewhere, or did you find these values by like trial and error?

> Actually, I think I slightly remember. GTA IV was particularly obnoxious because the ingame 0,0 isn't in the middle of the map.

Yeah, I've noticed that when I was doing some GTA IV modding and wanted to draw some ingame stuff on the 6000x3800 image of the map and had to come up with some "ingame-coordinates-to-pixel-position" formula.

The main problem being, finding the correct scaling, since even if I use the correct offsets to find the ingame 0,0 on the image, the further things are away from that point, the more misaligned they become.

Anyway, thanks for the explanation, maybe you can figure out a way to get the bounds working for the other maps too, so they will work flawlessly.

**Comment by interactive-game-maps** [comment]
> So does Leaflet calculate that value based on the tilesize at a certain zoom level, is it defined somewhere, or did you find these values by like trial and error?

I just looked at the resulting map back then.

But this isn't a thing of leaflet but of the converting python script. It seems rather easy to calculate from the original file size: https://github.com/commenthol/gdal2tiles-leaflet/blob/master/gdal2tiles.py#L546-L548
The script also puts the original image size into the `tilemapresource.xml` inside `<BoundingBox />`.
Together with a fallback to `256`, the maximum possible, we should be good to go.

Do you want to have a shot at this?

**Comment by interactive-game-maps** [comment]
Fixed by 91c7d4dc0fb12d1f95623091a940cb61582e8f67
